File: twitterDesign.java
class Tweet {
    int id;
    int timestamp;
        
    public Tweet(int id, int timestamp) {
        this.id = id;
        this.timestamp = timestamp;
    }
}

class Twitter {

    HashMap<Integer, HashSet<Integer>> follows = null;
    HashMap<Integer, List<Tweet>> userTweets = null;
    PriorityQueue<Tweet> mostRecentTweets = null;
    int timestamp;
    
    public Twitter() {
        timestamp = 0;
        follows = new HashMap<>();
        userTweets = new HashMap<>();
        mostRecentTweets = new PriorityQueue<Tweet>((t1, t2) -> t1.timestamp - t2.timestamp);
    }
    
    // TC: O(1)
    public void postTweet(int userId, int tweetId) {
        ++timestamp;
        if (!userTweets.containsKey(userId)) {
            userTweets.put(userId, new LinkedList<>());
        }
        userTweets.get(userId).add(new Tweet(tweetId, timestamp));
        // TODO: Cap at 10
    }
    
    // TC: O(n * m * log 10) => O(n * m) where m = max number of tweets.
    public List<Integer> getNewsFeed(int userId) {
        
        mostRecentTweets.clear();
        
        // We need to add tweets of the user itself
        addToMostRecentTweets(userId);
        
        // Add tweets of the user's followers
        HashSet<Integer> followers = follows.get(userId);
        if (followers != null) { 
            for (int follower : followers) {
                addToMostRecentTweets(follower);
            }
        }
        
        List<Integer> result = new ArrayList<>();
        while (!mostRecentTweets.isEmpty()) {
            result.add(0, mostRecentTweets.poll().id);
        }
        return result;
    }
    
    // TC: O(1)
    public void follow(int followerId, int followeeId) {
        if (!follows.containsKey(followerId)) {
            follows.put(followerId, new HashSet<>());
        }
        follows.get(followerId).add(followeeId);
    }
    
    // TC: O(1)
    public void unfollow(int followerId, int followeeId) {
        if (follows.containsKey(followerId)) {
            follows.get(followerId).remove(followeeId);
        }
    }
    
    private void addToMostRecentTweets(int userId) {
        List<Tweet> tweets = userTweets.get(userId);
        
        if (tweets != null) {
            for (Tweet t : tweets) {
                mostRecentTweets.offer(t);
                if (mostRecentTweets.size() > 10) {
                    mostRecentTweets.poll();
                }
            }
        }
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: TwitterDesign.cs
# TC:
# postTweet: O(1)
# getNewsFeed: O(n)
# follow: O(1)
# unfollow: O(1)

public class Twitter
{
    Dictionary<int, IList<(int TweetId, int Time)>> userMap;
    Dictionary<int, HashSet<int>> followerMap;

    private int time;

    public Twitter()
    {
        time = 0;
        userMap = new Dictionary<int, IList<(int TweetId, int Time)>>();
        followerMap = new Dictionary<int, HashSet<int>>();
    }

    public void PostTweet(int userId, int tweetId)
    {
        if (!userMap.ContainsKey(userId))
        {
            userMap[userId] = new List<(int TweetId, int Time)>();
        }
        userMap[userId].Add((tweetId, time));
        time++;
    }

    public IList<int> GetNewsFeed(int userId)
    {
        var tweetIdList = new List<int>();
        var userIdListToRetrieve = new List<int>();
        var pq = new PriorityQueue<int, int>(Comparer<int>.Create((x, y) => y - x));

        if (!userMap.ContainsKey(userId)) return tweetIdList;

        if (followerMap.ContainsKey(userId))
        {
            userIdListToRetrieve.AddRange(followerMap[userId]);
        }
        userIdListToRetrieve.Add(userId);

        foreach (var id in userIdListToRetrieve)
        {
            for (int i = Math.Max(0, userMap[id].Count - 10); i < userMap[id].Count; ++i)
            {
                pq.Enqueue(userMap[id][i].TweetId, userMap[id][i].Time);
            }
        }

        while (tweetIdList.Count < 10 && pq.Count > 0)
        {
            tweetIdList.Add(pq.Dequeue());
        }
        return tweetIdList;
    }

    public void Follow(int followerId, int followeeId)
    {
        if (!userMap.ContainsKey(followerId))
        {
            userMap[followerId] = new List<(int TweetId, int Time)>();
        }
        if (!userMap.ContainsKey(followeeId))
        {
            userMap[followeeId] = new List<(int TweetId, int Time)>();
        }
        if (!followerMap.ContainsKey(followerId))
        {
            followerMap[followerId] = new HashSet<int>();
        }

        followerMap[followerId].Add(followeeId);
    }

    public void Unfollow(int followerId, int followeeId)
    {
        if (followerMap.ContainsKey(followerId))
        {
            followerMap[followerId].Remove(followeeId);
        }
    }
}
/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.PostTweet(userId,tweetId);
 * IList<int> param_2 = obj.GetNewsFeed(userId);
 * obj.Follow(followerId,followeeId);
 * obj.Unfollow(followerId,followeeId);
 */
